### PR TITLE
PX shouldn't modify attrs controlled by template

### PR DESCRIPTION
**Pull request by nicolaskruchten**
This supersedes #1863

### CODE REVIEW

**Comment by emmanuelle** [comment]
Thanks it looks good to me. I'm glad that creating a `go.layout.Template` object helps to remove some of the `try/except`. As we discussed previously, adding a bunch of unit tests here would be great.

**Comment by nicolaskruchten** [comment]
This now supersedes #1871 also

**Comment by nicolaskruchten** [comment]
@emmanuelle ready for re-review! I added a nice little test suite.

File: packages/python/plotly/plotly/tests/test_core/test_px/test_px.py [hunk 1]
```diff
@@ -51,3 +51,111 @@ def test_custom_data_scatter():
...
+    assert fig.layout.template == pio.templates[pio.templates.default]
+
+    # respect changes to defaults
+    pio.templates.default = "seaborn"
+    fig = px.scatter()
+    assert fig.layout.template == pio.templates["seaborn"]
+
+    # special px-level defaults over pio defaults
+    pio.templates.default = "seaborn"
+    px.defaults.template = "ggplot2"
```
Comment (emmanuelle) [inline packages/python/plotly/plotly/tests/test_core/test_px/test_px.py]: we should document this in https://plot.ly/python/templates/, I did not know this was possible

**Comment by emmanuelle** [comment]
Thanks for adding this awesome test suite! We might want to merge #1877 first and rebase to have a green CI and benefit from Percy visual inspection, but this looks really good.

**Comment by emmanuelle** [comment]
:dancer: once the CI is green

**Pull request merged**